activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -162,34 +162,37 @@ def initialize(pool_config)
         @threads_blocking_new_connections = 0
 
         @available = ConnectionLeasingQueue.new self
-
-        @lock_thread = false
+        @pinned_connection = nil
 
         @async_executor = build_async_executor
 
         @reaper = Reaper.new(self, db_config.reaping_frequency)
         @reaper.run
       end
 
-      def lock_thread=(lock_thread)
-        if lock_thread
-          @lock_thread = ActiveSupport::IsolatedExecutionState.context
-        else
-          @lock_thread = nil
-        end
-
-        if (active_connection = @thread_cached_conns[connection_cache_key(current_thread)])
-          active_connection.lock_thread = @lock_thread
-        end
-      end
-
       # Retrieve the connection associated with the current thread, or call
       # #checkout to obtain one if necessary.
       #
       # #connection can be called any number of times; the connection is
       # held in a cache keyed by a thread.
       def connection
-        @thread_cached_conns[connection_cache_key(current_thread)] ||= checkout
+        @thread_cached_conns[connection_cache_key(ActiveSupport::IsolatedExecutionState.context)] ||= checkout
+      end
+
+      def pin_connection! # :nodoc:
+        raise "There is already a pinned connection" if @pinned_connection
+
+        @pinned_connection = (@thread_cached_conns[connection_cache_key(ActiveSupport::IsolatedExecutionState.context)] || checkout)
+        yield @pinned_connection
+      end
+
+      def unpin_connection! # :nodoc:
+        raise "There isn't a pinned connection" unless @pinned_connection
+
+        yield @pinned_connection
+        old_connection = @pinned_connection
+        @pinned_connection = nil
+        checkin(old_connection)
```

Comment:
Seems surprising if `pin_connection!` adopted its caller's already-cached connection

Reply:
Yeah, this also wasn't there initially, but I has to do it to make some tests pass. I think some tests do activate a connection before transactional fixtures kick in, but that's something I need to dig a bit deeper in.